x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/SearchAfterDocumentsIterator.java

```diff
@@ -0,0 +1,171 @@
+ * An iterator useful to fetch a large number of documents of type T
+ * and iterate through them in batches of 10,000.
+ *
+ * In terms of functionality this is very similar to {@link BatchedDocumentsIterator}
+ * the difference being that this uses search after rather than scroll.
+ *
+ * Search after has the advantage that the scroll context does not have to be kept
+ * alive so if processing each batch takes a long time search after should be
+ * preferred to scroll.
+ */
+public abstract class SearchAfterDocumentsIterator<T> implements BatchedIterator<T> {
+
+    private static final int BATCH_SIZE = 10_000;
+
+    private final OriginSettingClient client;
+    private final String index;
+    private volatile long count;
+    private volatile long totalHits;
+
+    protected SearchAfterDocumentsIterator(OriginSettingClient client, String index) {
+        this.client = Objects.requireNonNull(client);
+        this.index = Objects.requireNonNull(index);
+        this.totalHits = -1;
+        this.count = 0;
+    }
+
+    /**
+     * Returns {@code true} if the iteration has more elements or
+     * no searches have been been run and it is unknown if there is a next.
+     *
+     * @return {@code true} if the iteration has more elements or the first
+     * search has not been run
+     */
+    @Override
+    public boolean hasNext() {
+        return count != totalHits;
+    }
+
+    /**
+     * The first time next() is called, the search will be performed and the first
+     * batch will be returned. Any subsequent call will return the following batches.
+     * <p>
+     * Note that in some implementations it is possible that when there are no
+     * results at all, the first time this method is called an empty {@code Deque} is returned.
+     *
+     * @return a {@code Deque} with the next batch of documents
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    @Override
+    public Deque<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        SearchResponse searchResponse;
+        if (totalHits == -1) {
+            searchResponse = initSearch();
+        } else {
+            searchResponse = doSearch(searchAfterFields());
+        }
+        return mapHits(searchResponse);
+    }
+
+    private SearchResponse initSearch() {
+        SearchResponse searchResponse = doSearch(null);
+        totalHits = searchResponse.getHits().getTotalHits().value;
+        return searchResponse;
+    }
+
+    private SearchResponse doSearch(Object [] searchAfterValues) {
+        SearchRequest searchRequest = new SearchRequest(index);
+        searchRequest.indicesOptions(MlIndicesUtils.addIgnoreUnavailable(SearchRequest.DEFAULT_INDICES_OPTIONS));
+        SearchSourceBuilder sourceBuilder = (new SearchSourceBuilder()
+            .size(BATCH_SIZE)
+            .query(getQuery())
+            .fetchSource(shouldFetchSource())
+            .trackTotalHits(true)
```

Comment:
We only need total hits for the first query. Would be good to have this be `false` once we have the total hits recorded.